src/github_runner_image_builder/store.py

```diff
@@ -0,0 +1,115 @@
+# Copyright 2024 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Module for uploading images to shareable storage."""
+
+import logging
+from pathlib import Path
+from typing import cast
+
+import openstack
+import openstack.connection
+import openstack.exceptions
+from openstack.image.v2.image import Image
+
+from github_runner_image_builder.errors import GetImageError, OpenstackError, UploadImageError
+
+logger = logging.getLogger(__name__)
+
+
+def upload_image(cloud_name: str, image_name: str, image_path: Path, keep_revisions: int) -> str:
+    """Upload image to openstack glance.
+
+    Args:
+        cloud_name: The Openstack cloud to use from clouds.yaml.
+        image_name: The image name to upload as.
+        image_path: The path to image to upload.
+        keep_revisions: The number of revisions to keep for an image.
+
+    Raises:
+        UploadImageError: If there was an error uploading the image to Openstack Glance.
+
+    Returns:
+        The created image ID.
+    """
+    with openstack.connect(cloud=cloud_name) as connection:
+        try:
+            image: Image = connection.create_image(
+                name=image_name,
+                filename=str(image_path),
+                allow_duplicates=True,
+                wait=True,
+            )
+            _prune_old_images(
+                connection=connection, image_name=image_name, num_revisions=keep_revisions
+            )
+            return image.id
+        except openstack.exceptions.OpenStackCloudException as exc:
+            raise UploadImageError from exc
+
+
+def _prune_old_images(
+    connection: openstack.connection.Connection, image_name: str, num_revisions: int
+) -> None:
+    """Remove old images outside of number of revisions to keep.
+
+    Args:
+        connection: The connected openstack cloud instance.
+        image_name: The image name to search for.
+        num_revisions: The number of revisions to keep.
+    """
+    images = _get_sorted_images_by_created_at(connection=connection, image_name=image_name)
```

Comment:
`_get_sorted_images_by_created_at` might throw an error, I'm wondering if we shouldn't catch it and just return as the prune shouldn't interrupt the processing

Reply:
I think prune should interrupt - it is a feature `--keep-revisions` and also it is quite frequent that users will not notice warnings/logs and in case of the charm, it has a possibility to continue stacking up the images in Glance.